core/unix/rseq_linux.c

```diff
@@ -0,0 +1,377 @@
+
+typedef struct _rseq_region_t {
+    app_pc start;
+    app_pc end;
+    app_pc handler;
+    /* XXX i#2350: Additional information will be added here for eliminating
+     * some existing limitations.
+     */
+} rseq_region_t;
+
+/* vmvector callbacks */
+static void
+rseq_area_free(void *data)
+{
+    HEAP_TYPE_FREE(GLOBAL_DCONTEXT, data, rseq_region_t, ACCT_VMAREAS, PROTECTED);
+}
+
+static void *
+rseq_area_dup(void *data)
+{
+    rseq_region_t *src = (rseq_region_t *)data;
+    rseq_region_t *dst =
+        HEAP_TYPE_ALLOC(GLOBAL_DCONTEXT, rseq_region_t, ACCT_VMAREAS, PROTECTED);
+    ASSERT(src != NULL);
+    *dst = *src;
+    return dst;
+}
+
+void
+d_r_rseq_init(void)
+{
+    VMVECTOR_ALLOC_VECTOR(d_r_rseq_areas, GLOBAL_DCONTEXT,
+                          VECTOR_SHARED | VECTOR_NEVER_MERGE, rseq_areas);
+    vmvector_set_callbacks(d_r_rseq_areas, rseq_area_free, rseq_area_dup, NULL, NULL);
+    if (rseq_is_registered_for_current_thread())
+        rseq_enabled = true;
+}
+
+void
+d_r_rseq_exit(void)
+{
+    vmvector_delete_vector(GLOBAL_DCONTEXT, d_r_rseq_areas);
+    DELETE_LOCK(rseq_trigger_lock);
+}
+
+bool
+rseq_get_region_info(app_pc pc, app_pc *start OUT, app_pc *end OUT, app_pc *handler OUT)
+{
+    rseq_region_t *info;
+    if (!vmvector_lookup_data(d_r_rseq_areas, pc, start, end, (void **)&info))
+        return false;
+    if (handler != NULL)
+        *handler = info->handler;
+    return true;
+}
+
+bool
+rseq_is_registered_for_current_thread(void)
+{
+    /* Unfortunately there's no way to query the current rseq struct.
+     * For 64-bit we can pass a kernel address and look for EFAULT
+     * vs EINVAL, but there is no kernel address for 32-bit.
+     * So we try to perform a legitimate registration.
+     */
+    struct rseq test_rseq = {};
+    int res = dynamorio_syscall(SYS_rseq, 4, &test_rseq, sizeof(test_rseq), 0, 0);
+    if (res == -EINVAL) /* Our struct != registered struct. */
+        return true;
+    if (res == -ENOSYS)
+        return false;
+    ASSERT(res == 0); /* If not, the struct size or sthg changed! */
+    if (dynamorio_syscall(SYS_rseq, 4, &test_rseq, sizeof(test_rseq),
+                          RSEQ_FLAG_UNREGISTER, 0) != 0) {
+        ASSERT_NOT_REACHED();
+    }
+    return false;
+}
+
+static void
+rseq_process_entry(struct rseq_cs *entry, ssize_t load_offs)
+{
+    LOG(GLOBAL, LOG_LOADER, 2,
+        "Found rseq region: ver=%u; flags=%u; start=" PFX "; end=" PFX "; abort=" PFX
+        "\n",
+        entry->version, entry->flags, entry->start_ip + load_offs,
+        entry->start_ip + entry->post_commit_offset + load_offs,
+        entry->abort_ip + load_offs);
+    rseq_region_t *info =
+        HEAP_TYPE_ALLOC(GLOBAL_DCONTEXT, rseq_region_t, ACCT_VMAREAS, PROTECTED);
+    info->start = (app_pc)(ptr_uint_t)entry->start_ip + load_offs;
+    info->end = info->start + entry->post_commit_offset;
+    info->handler = (app_pc)(ptr_uint_t)entry->abort_ip + load_offs;
+#if 0 // NOCHECK
```

Comment:
Something better than if 0?

Reply:
This is my bad -- I extracted this refactoring out of a branch implementing invoke-start-instead-of-handler for better incremental commits, but failed to properly remove all of the in-progress code from that.  Removing.